### ISSUE TITLE
Incorrect Reading statistics > Reading progress after midnight

### ISSUE THREAD

**Issue by charleshan** [koreader/koreader#12371]
* KOReader version: 2024.07-63
* Device: Boox Poke 3

#### Issue
The `Reading progress` numbers are different from `Today's timeline` and it looks like `Reading progress` is the one that's incorrect.

`Reading progress` Session Pages 0, Time 00:00 (I think this is incorrect. I just read a chapter or so)
`Reading progress` Today Pages 76, Time 02:04

`Today's timeline` 00:39:09 (25 pages)
Yesterday 1:24:23 (51 pages read after midnight)

I believe this is because `Daily timeline starts at` setting isn't being applied to `Reading progress`. Maybe this hasn't been implemented yet but I think we can agree that it's not working as expected.

**Comment by poire-z** [comment]
> Maybe this hasn't been implemented yet

Probably.

> but I think we can agree that it's not working as expected.

May be hard to implement, dunno - there are other things that are computed differerently - ie. directly by a SQL query - where we can't easily make use of the "Daily timeline starts at HH:00" (if that's what you're really talking about, and you have changed it from the default of 00:00 ? if not, it may be some UTC/localtime shift not accounted).

**Comment by charleshan** [comment]
Yes I have changed `Daily timeline starts at` to 04:00.

I haven't looked into the code but there must be something in place that allows `Calendar view` and `Today's timeline` to work properly.

**Comment by charleshan** [comment]
Related: https://github.com/koreader/koreader/issues/2978

This was added for `Calendar view` and `Today's timeline` but not for `Reading progress`

**Comment by remizik** [comment]
+1 on this, I came here to post this.
(Device is a Kobo claraBW on KR version 2024.11

**Comment by ironhak** [comment]
+1

**Comment by OverkillGuy** [comment]
> This was added for `Calendar view` and `Today's timeline` but not for `Reading progress`

Indeed! This would explain part of #10752 too regarding reading progress.
I'm personally quite interested in "Reading progress" being accurate per that "daily timeline start"!

### Quick Investigation
Tracking down the code for it, I see that the "bad" code of "Reading progress" is: https://github.com/koreader/koreader/blob/fe3775b077f598f2dd85686d928fa5f1d1cc813e/plugins/statistics.koplugin/main.lua#L2162

compared to the "good" code referencing the offset: 
https://github.com/koreader/koreader/blob/fe3775b077f598f2dd85686d928fa5f1d1cc813e/plugins/statistics.koplugin/main.lua#L2826
and then used against `start_time`:
https://github.com/koreader/koreader/blob/fe3775b077f598f2dd85686d928fa5f1d1cc813e/plugins/statistics.koplugin/main.lua#L2834.

I'll try to make a pull request adopting that offset thing.

But would the maintainers be happy to take that in? Reading other threads like #10752 I wasn't sure if this made consensus?

**Comment by poire-z** [comment]
Why not if it bothers some, as long as there's an option to _not lie_ about what a day is :)
(Either making the existing `[ ] Also use in calendar view` more general, or adding a `[ ] Also use everywhere else`.)
Are all the "features" missing that contained in the statistics.koplugin ? Or are there other stuff elsewhere depending on it? Might be a good idea to make that day start shift available via a self.ui.statistics methods so the new fancy "Home/UI" external plugins showing "today stuff" can also fetch it (dunno how they do that currently).

And if you are getting familiar with all that, maybe you'll have some insight on this puzzling #14122...

**Comment by OverkillGuy** [comment]
> Why not if it bothers some, as long as there's an option to *not lie* about what a day is :) 

Fair, thank you. I have a [first try for the basic patch](https://github.com/koreader/koreader/compare/master...OverkillGuy:koreader:readprogress-startat?expand=1), which I need to test on my device.

I need to make a dummy stats database script that exhibits the problem, for testing.

> (Either making the existing `[ ] Also use in calendar view` more general, or adding a `[ ] Also use everywhere else`.)

It feels right, yes, something like "Also use in every other views"? I'll look into that after I make the basic thing work (and check other such views too for similar issue? )

> Are all the "features" missing that contained in the statistics.koplugin? Or are there other stuff elsewhere depending on it?

I was doing first round of checks, and yeah it looks like it: all uses of `page_stat` table are in the `main.lua` of the statistics plugin, except one that looks like a copy of ours, but from the exporter plugin for XMNote:

    plugins/exporter.koplugin/target/xmnote.lua
    89:            FROM   page_stat

and one from a db migration file?

So yeah, sounds like a positive, all in one place-ish except for one (which can likely be collapsed).

> Might be a good idea to make that day start shift available via a self.ui.statistics methods so the new fancy "Home/UI" external plugins showing "today stuff" can also fetch it (dunno how they do that currently).

Not sure how to do that, and for now I just want to finish the first pass on the feature to get a feel for the workflow, but I would be happy to learn in a second phase!

> And if you are getting familiar with all that, maybe you'll have some insight on this puzzling [#14122](https://github.com/koreader/koreader/issues/14122)&hellip;

Same, would love to have a look once I'm feeling more confident, especially from having a sample stats DB to play with.

**Comment by OverkillGuy** [comment]
## Test Data!

Just managed to get this sample script to generate a plausible statistics db:

[test_stats.sql](https://github.com/user-attachments/files/27575723/test_stats.sql)

Resulting `statistics.sqlite3` = 
[test_stats.sqlite3.gz](https://github.com/user-attachments/files/27575771/test_stats.sqlite3.gz) (had to gzip it for github, otherwise fails upload, sorry).

### Contents
A single book of 1000 pages, pages 1, 2, 3 ... 200 read for 60 seconds each, page 1 started at 23:00 UTC, page 2 at 23:01, and so on for 200 pages = 3h 20 mins of reading, so the last page is read at ~2:20 AM:

```sql
-- First page read:
SELECT min(strftime('%Y-%m-%dT%H:%M:%SZ', start_time, 'unixepoch')) FROM page_stat;
-- 2026-05-10T23:00:00Z

-- Last page read:
SELECT max(strftime('%Y-%m-%dT%H:%M:%SZ', start_time, 'unixepoch')) FROM page_stat;
-- 2026-05-11T02:19:00Z
```

Regenerate the db:
```shell
sqlite3 test_stats.sqlite3 < test_stats.sql
```
and copy that under the name `statistics.sqlite3` in the right folder.

We should be able to use this file as reference for testing both the daily timeline shifting _disabled_ case, and the case where _all the reading is in the past day_ (using my own preference of daily timeline = 5AM).

If we're lucky, we'll be able to fiddle with this sample database to explore the other timezone-adjacent bug next, and maybe even define actual automated regression tests based on it?

I'll play with this dataset, of course, but I was excited to have data that may help others test this too.
